### PR TITLE
fix(core): set default ariaLabel to Message Strip for screen reader a11y

### PR DESCRIPTION
**Pull request by kavya-b**
#### Please provide a link to the associated issue.
Closes SAP/fundamental-ngx#3123

#### Please provide a brief summary of this pull request.
This PR adds the `ariaLabel` property to some of the examples for the screen readers to properly convey the message strip content. PR also handles logic for when `ariaLabel` property is not specified, the library will set the text of the message strip as the aria-label.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [n/a] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [x] Run npm run build-pack-library and test in external application

Documentation checklist:
- [n/a] update `README.md`
- [n/a] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

**Review by platon-rov — APPROVED** [review]

**Review by DeepakSap14 — APPROVED** [review]

**Review by InnaAtanasova — APPROVED** [review]

**Review by InnaAtanasova — APPROVED** [review]

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
Would be nice if the ariaLabel was set automatically - this can be done by giving the <ng-content> for the text a template variable, accessing it via ContentChild in the component, and then if the developer has not explicitly set the ariaLabel via input then to set it in AfterContentInit

**Comment by kavya-b** [comment]
> Would be nice if the ariaLabel was set automatically - this can be done by giving the for the text a template variable, accessing it via ContentChild in the component, and then if the developer has not explicitly set the ariaLabel via input then to set it in AfterContentInit

@mikerodonnell89 using ElementRef reference to set the aria-label value to the message strip text in AfterViewInit. ContentChild was not holding the reference correctly, so used ViewChild instead. Please take a look. 

@InnaAtanasova @DeepakSap14 @platon-rov please re-review since code was changed after your approvals.

**Review by platon-rov — APPROVED** [review]

**Review by InnaAtanasova — COMMENTED** [review]
Now that ariaLabel is set by default (to the innerText) I think only 1 example can have a custom ariaLabel and the rest can be as they were before. The user might get the impression that they have to always provide `ariaLabel`. Also, I think it's a good idea to mention this functionality in the documentation (that if the aria label is omitted it will be set to the content).

File: libs/core/src/lib/message-strip/message-strip.component.ts
Comment (InnaAtanasova) [inline libs/core/src/lib/message-strip/message-strip.component.ts:113]: Just a question. What will happen if the user passes a more complex content like text+button or text+link or even another component?

**Comment by mikerodonnell89** [comment]
Testing the current build at https://fundamental-ngx.netlify.com/ with Chrome's screenreader, it is actually reading the message strips correctly when I click them, and with the deployment preview on this PR, the screenreader is reading the text from the message strip twice. @kavya-b what screen reader are you testing with?

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
Testing the current build at https://fundamental-ngx.netlify.com/ with Chrome's screenreader, it is actually reading the message strips correctly when I click them, and with the deployment preview on this PR, the screenreader is reading the text from the message strip twice. @kavya-b what screen reader are you testing with?

File: libs/core/src/lib/message-strip/message-strip.component.ts
Comment (kavya-b) [reply libs/core/src/lib/message-strip/message-strip.component.ts:113]: hmm, not sure about this. How do you think we should handle this then?

**Comment by kavya-b** [comment]
> Testing the current build at https://fundamental-ngx.netlify.com/ with Chrome's screenreader, it is actually reading the message strips correctly when I click them, and with the deployment preview on this PR, the screenreader is reading the text from the message strip twice. @kavya-b what screen reader are you testing with?

I am testing it with Voiceover on Mac.. what I understood from the issue was the screen reader was earlier only reading the button(‘Dismiss’) on entering the strip, but it should also read the text. In VoiceOver, we can explicitly navigate through Ctrl+option left/right arrows and it reads the text(without needing any aria label which is probably why it works on the latest netlify too), but apparently the issue requires that it be read out on entering the whole message strip. So now with the aria label, it reads the text on entering the strip through Ctrl+option left/right arrows(in Mac) and then it reads the button and text as well explicitly,  which is why it seems to be repeating twice.

File: libs/core/src/lib/message-strip/message-strip.component.ts
Comment (InnaAtanasova) [reply libs/core/src/lib/message-strip/message-strip.component.ts:113]: Screen Shot 2021-05-26 at 1 31 01 PM

I tested this with adding a button to the text. It takes the content of the button and append it to the other text.

Comment (InnaAtanasova) [reply libs/core/src/lib/message-strip/message-strip.component.ts:113]: @mikerodonnell89 FYI

**Comment by kavya-b** [comment]
Had a call with the a11y team member alongwith Puru, and this is basically what they are expecting from a message strip to read out:
- an invisible text for the semantic icon at the start (Message Strip Error Closable)
- the text(There are no values distributed to characteristic Location)
- The type of message strip(Error Message Strip Close Information Bar)
- the close button(Error Message Strip Close Button)
Screenshot 2021-05-27 at 10 22 55 AM
(the image is an example from a UI5 project that a11y member showed us)

we may need to place some invisible span tags to read out these information @InnaAtanasova @mikerodonnell89 
cc: @puru-hk @manjunathanagaraj

**Review by Betrozov — APPROVED** [review]

**Comment by kavya-b** [comment]
@InnaAtanasova @mikerodonnell89  made the following changes:
- [text that is already visible should not have aria-label on it](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute), therefore the previous approach of attaching aria-label directly to the text element by reading its contents was wrong. The suggested approach is to use aria-labelledby on the parent(`fd-message-strip`) to read out the content through the id. 
- so now what we do is we attach the `fd-message-strip__text` class to the host message strip’s `aria-labelledby`.
since we accept `aria-labelledby` from user as well, if provided, we can add both these ids to the host aria-labelledby. 
- we have 2 invisible spans now: one that reads out the strip information with type/icon info, and another that holds the custom aria label that user may provide through `ariaLabel` property. With the latter, we want the screen reader to read out the custom aria label and not what is provided from  aria-labelledby. An example of this is the 2nd message strip of the last example.
image

On voiceover, it reads only once for me now, please check.

File: libs/core/src/lib/message-strip/message-strip.component.html
Comment (mikerodonnell89) [inline libs/core/src/lib/message-strip/message-strip.component.html:1]: I do not think this invisible element is necessary, this aria-label can be placed on the <fd-message-strip> host element

File: libs/core/src/lib/message-strip/message-strip.component.html
Comment (mikerodonnell89) [inline libs/core/src/lib/message-strip/message-strip.component.html:8]: after removing the invisible element and moving its aria-label to the host, I think we can get rid of this aria-hidden="true"

File: libs/core/src/lib/message-strip/message-strip.component.ts
Comment (mikerodonnell89) [inline libs/core/src/lib/message-strip/message-strip.component.ts:167]: This is never read on the chrome screenreader

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]

File: libs/core/src/lib/message-strip/message-strip.component.html
Comment (kavya-b) [reply libs/core/src/lib/message-strip/message-strip.component.html]: have followed your approach and tested with chrome screen reader. Thank you!
Will get in touch with a11y member to test on JAWS as well. Seems like voiceover is not a good way to test screen reader stuff!

File: libs/core/src/lib/message-strip/message-strip.component.ts
Comment (kavya-b) [reply libs/core/src/lib/message-strip/message-strip.component.ts]: with [Mike's approach](https://github.com/SAP/fundamental-ngx/pull/5458#discussion_r641021336) the `ariaLabel` itself will now take the custom message.

File: libs/core/src/lib/message-strip/message-strip.component.ts
Comment (kavya-b) [reply libs/core/src/lib/message-strip/message-strip.component.ts]: there was no need of these fields anymore, so have removed it.

**Review by InnaAtanasova — APPROVED** [review]

**Review by Betrozov — APPROVED** [review]

**Comment by kavya-b** [comment]
This PR is blocked until we can confirm whether to add focus to Message Strip or not, in order for screen reader to read out correctly in JAWS(Windows). Please check the comments in the issue #3123 and provide your inputs. @InnaAtanasova @droshev

**Comment by kavya-b** [comment]
awaiting a11y document for Message Strip from a11y team for reference.

**Comment by kavya-b** [comment]
Closing this PR for now. Will reopen it if the issue gets reopened again.

**Comment by Veerakarthick** [comment]
Member
[Veerakarthick](https://github.com/Veerakarthick) commented [26 minutes ago](https://github.com/SAP/fundamental-ngx/issues/3123#issuecomment-1074673441)[](https://user-images.githubusercontent.com/37998745/159398889-f5db029b-dce8-430f-8188-596f461fb615.png)
Hello Colleague,

Retest has been taken reported issue still persist the same so please check and do the needful.
Description: When the message strips are accessed, the information inside the message strips are not conveyed (screen reader)

Expected: The content inside the message strips should be conveyed properly when the message strips are accessed(screen reader)

Screenshot:
image

Retest envornment:
URL: https://deploy-preview-5458--fundamental-ngx.netlify.app/
JAWS 2021

Regards,
Karthikeyan.c

**Comment by Karthickc1994** [comment]
Hello Colleague,

Retest has been taken with the below environment test results are below:

Issue: When the message strip is accessed the information within the message strip is not read out.
Reading made as:
Close Button
To activate press spacebar.
Close Button
To activate press spacebar.
Close Button
To activate press spacebar.
 Show Code Button
To activate press spacebar.

Screen shot:
image

Expected behavior:
when the message strip is accessed the information within the message strip should read out by the screen reader.

Retest Environment:
URL:[https://deploy-preview-5458--fundamental-ngx.netlify.app](https://deploy-preview-5458--fundamental-ngx.netlify.app/?utm_source=github&utm_campaign=bot_dp)
Browser: Version 101.0.4951.54 (Official Build) (64-bit)
Screen Reader: JAWS 2022.2204.20 ILM

Regards,
Karthikeyan.c

**Pull request closed without merging**
